### PR TITLE
[Storage_ipc] Option II: Provides IPC extensions for 3rd devices. 

### LINKED ISSUE #124902: Add support for IPC features for PrivateUse devices
### 🚀 The feature, motivation and pitch

Currently, PyTorch only supports inter-process communication (IPC) on CUDA devices. I'm looking into doing IPC on 3rd party devices and hope some abstract code can be added to the PyTorch community.
So I want to ask if this feature is considered

### Alternatives

I plan to provide underlying support through PrivateUse1Hooks and abstract a PrivateUse1IPCTypes method, and provide PrivateUse1 device support for the _share_cuda_, _new_shared_cuda, and _release_ipc_counter methods in the python Storage class.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mengpenghui**
I am working on IPC-related interfaces to support third-party devices. 
This is another option for IPC support for third party devices.

1. The share_cuda_ interface is changed to share_device_
2. The backend uses getAccelerator to determine the device and distribute different devices.
3. Add the share_device implementation of third-party devices to PrivateUse1HooksInterface

In this way, third-party devices and CUDA will use the same interface, and the device distinction is hidden in the backend. Users also need to manually set the device type (although these are internal interfaces).

another plan: https://github.com/pytorch/pytorch/pull/125122

Do you think which is feasible? Looking forward to your suggestions
Fixes #124902

cc @albanD

### CODE REVIEW

File: torch/csrc/StorageSharing.cpp [hunk 1]
```diff
@@ -651,18 +652,77 @@ PyObject* THPStorage_isShared(PyObject* self, PyObject* noargs) {
   }
 }
 
+static PyObject* THPStorage_shareDecice(PyObject* self, PyObject* noargs) {
```
Comment (legionGIT) [inline torch/csrc/StorageSharing.cpp:655]: shareDecice should be shareDevice

**Review by albanD — COMMENTED** [review]
Thanks for the PR, that sounds good in principle but I think the implementation needs a bit of work for two things:
- the aten level hooks should not take as input and return PyObject*. This is the c++ layer and should not deal with arg-parsing in there. I think in this case, you want to refactor the arg-parsing from the cuda code to be shared. And then call into a c++ only hook.
- The hook should be generic, I think the AcceleratorHookInterface is the right place to do that. This way it can be shared by CUDA/PrivateUse1/others
- The python layer should be generic using acceleratorhooks so that you don't need if/else for each device.

File: c10/core/StorageImpl.h [hunk 2]
```diff
@@ -294,7 +294,7 @@ struct C10_API StorageImpl : public c10::intrusive_ptr_target {
   bool resizable_;
   // Identifies that Storage was received from another process and doesn't have
   // local to process cuda memory allocation
-  bool received_cuda_;
+  bool received_device_;
```
[the remark this answers is not in the document]
Comment (mengpenghui) [reply c10/core/StorageImpl.h:297]: update completed

**Comment by mengpenghui** [comment]
@albanD 
Thank you for your review. I have refactored the code according to the above suggestions, mainly including:
1. Put IPC related methods into AcceleratorHookInterface, making it a common method for CUDA/PrivateUse1/others backend.
2. Process arg-parsing in StorageSharing.cpp, and match IPC methods of different devices through getAcceleratorHooksInterface.
3. Put IPC and cudart related codes into cudahooks.
4. Add inter-process communication processing of third-party devices in torch/multiprocessing/reductions.py.
Currently tested in test_multiprocessing.py.
Please review, FYI

File: c10/core/StorageImpl.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (mengpenghui) [reply c10/core/StorageImpl.h:297]: @albanD please review

**Pull request closed without merging**